### PR TITLE
[Optimized Implementation] Proposal to Fix flood HOM

### PR DESCRIPTION
**Pull request by JadingTsunami**
Gaps in flat rendering cause a HOM void in the GL renderer where the software renderer would bleed through the flat behind the void. This code is an initial attempt to remedy this (see #226). 

There are a few drawbacks:

* It pulls `P_Spec` calls into `R_Bsp` which is not ideal from an encapsulation perspective.
* A known corner case exists if an off-screen control sector is adjacent to the on-screen sector at a lower/higher sector height. This will cause the fake flat to take on the properties of the wrong sector. However, due to the nature of this effect, I have never seen the sector effect used this way and I suspect it may be rare to never.

Comment is invited as I believe there are many ways to fix this, and some may see advantages to one approach over others.

### CODE REVIEW

**Review by fabiangreffrath — COMMENTED** [review]
Thank you very much for taking a look at this! I cannot say much about the change itself as I haven't wrapped my head around the OpenGL renderer yet. But the code looks clean and reasonable. 

Remaining questions:
- Does this change also fix the rendering bug in E4M6, i.e. https://github.com/coelckers/prboom-plus/issues/173 ?
- Is it possible to preprocess and cache some of the calls to `P_FindLowestFloorSurrounding()` and `P_FindModelFloorSector()`?
- What exactly does `use_gl_stencil` mean and what happens if this is enabled? Will a different fix be applied? Does the user have any control over this variable?

**Comment by JadingTsunami** [comment]
Thanks for the review! Here are my thoughts...

> * Does this change also fix the rendering bug in E4M6, i.e. #173 ?

Unfortunately no, that one is caused by depth buffering and will require a different fix. Basically, the software renderer has no concept of a depth buffer, but it is an integral part of the OpenGL renderer. To mimic Doom's SW renderer, the GL renderer adds invisible "sky walls" to block the depth buffer (so walls "behind" the sky won't render at all). The problem is that the sky ceiling is *below* the top of the player's view height at that point, and those invisible sky walls block the marble walls from rendering. It's kind of a logical way to render the scene, but the Doom SW renderer does not always follow such logic... :)

> * Is it possible to preprocess and cache some of the calls to `P_FindLowestFloorSurrounding()` and `P_FindModelFloorSector()`?

Yes, in fact this is what the calls to `GetBestFake()` were doing, but I removed them after some consideration. Since it is a rare special effect (maybe 99% of maps do not use it), my reasoning was it's better to take a few extra cycles while rendering
 a rare "special effect" than add preprocessing time to every level.

But, I think perhaps I am thinking about it too deeply. :)

What changed my mind was, I would need to re-write parts of `GetBestFake` and add some new data structures. And adding more data structures, tracking and other stuff for this effect did not seem worth the tradeoff overall to me. But, if that is the preferred approach, I don't mind that way either.

> * What exactly does `use_gl_stencil` mean and what happens if this is enabled? Will a different fix be applied? Does the user have any control over this variable?

Yes, this can be configured in the main config file. It's whether or not the user's hardware supports stenciling. I think virtually all hardware supports it at this point. The stencil if check is actually legacy code that I left untouched.

In the SW renderer, if a sector has no bottom or top textures and is a raised floor or lowered ceiling, PrBoom will attempt to "bleed" that sector over to create the same effect as the SW renderer in such a case. There are two ways to do this: one that uses openGL stencils and one that does not. The code there is choosing the approach based on that support. The user could disable this in the config file, but I don't see why they would ever do this as the stencil approach should suffice.

**Comment by kraflab** [comment]
I'd vote to keep GetBestFake and make the changes there - at least, to see how they look.

**Comment by JadingTsunami** [comment]
> I'd vote to keep GetBestFake and make the changes there - at least, to see how they look.

Sure. I will mark this as WIP and give some time to optimize it.

**Comment by JadingTsunami** [comment]
Here's an updated example of what I'd do for preprocessing vs. calling p_spec in the renderer, with the new structure I envisioned. This also has a slight advantage that if the map author uses a control sector and correctly adds upper/lower textures, it should be handled correctly.

File: prboom2/src/r_bsp.c [hunk 1]
```diff
@@ -40,6 +40,7 @@
 #include "r_bsp.h" // cph - sanity checking
 #include "v_video.h"
 #include "lprintf.h"
+#include "p_spec.h"
```
Comment (fabiangreffrath) [inline prboom2/src/r_bsp.c:43]: Do we still need this here?

File: prboom2/src/gl_missingtexture.c [hunk 2]
```diff
@@ -111,6 +159,56 @@ static void gld_PrepareSectorSpecialEffects(void)
...
+
+    bleedsectors[source_idx].source = source;
+    bleedsectors[source_idx].ceiling = ceiling;
+
+    /* either register the proposed target since it is first,
+     * or check if the new proposed target is a better option
+     * and register it instead */
+
+    if ((bleedsectors[source_idx].target == NULL) ||
+             (bleedsectors[source_idx].target && ((ceiling && bleedsectors[source_idx].target->ceilingheight > target->ceilingheight) || (bleedsectors[source_idx].target->floorheight < target->floorheight))))
```
Comment (fabiangreffrath) [inline prboom2/src/gl_missingtexture.c:199]: Please break this line.

File: prboom2/src/gl_missingtexture.c [hunk 3]
```diff
@@ -111,6 +159,56 @@ static void gld_PrepareSectorSpecialEffects(void)
...
+        bleedsectors[source_idx].target = target;
+}
+
+sector_t* GetBestBleedSector(sector_t* source, int ceiling)
+{
+    int i;
+    for (i = 0; i < numbleedsectors; i++)
+        if (bleedsectors[i].source == source && bleedsectors[i].ceiling == ceiling)
```
Comment (fabiangreffrath) [inline prboom2/src/gl_missingtexture.c:207]: Why only ceiling?

File: prboom2/src/r_bsp.c [hunk 4]
```diff
@@ -664,39 +665,90 @@ static void R_Subsector(int num)
...
+                    {
+                        dummyceilingplane.height = tmpsec->ceilingheight;
+                        dummyceilingplane.lightlevel = tmpsec->lightlevel;
+                        dummyceilingplane.picnum = tmpsec->ceilingpic;
+                        ceilingplane = &dummyceilingplane;
+                    }
+                }
+            }
+
+            sector_t* tmpsec = NULL;
```
Comment (fabiangreffrath) [inline prboom2/src/r_bsp.c:705]: I think you could get away with only one tmpsec declaration at this point.

File: prboom2/src/gl_missingtexture.c [hunk 5]
```diff
@@ -111,6 +159,56 @@ static void gld_PrepareSectorSpecialEffects(void)
...
+    assert(target);
+
+    /* check whether the sector is processed already */
+    for (i = 0; i < numbleedsectors && source_idx == -1; i++)
+        if (bleedsectors[i].source == source && bleedsectors[i].ceiling == ceiling)
+            source_idx = i;
+    
+    if (source_idx == -1) {
+        /* allocate memory for new sector */
+        if (!bleedsectors) {
```
Comment (fabiangreffrath) [inline prboom2/src/gl_missingtexture.c:176]: There's an easier more generic solution that considers the fact that `realloc(NULL, ...)` is equal to `malloc(...)`. But that's mostly a matter of style.

File: prboom2/src/gl_missingtexture.c [hunk 6]
```diff
@@ -61,12 +61,22 @@ typedef struct
...
+    sector_t *source; // The sector to receive a fake bleed-through flat
+    sector_t *target; // The floor sector whose properties should be copied
+    int ceiling;      // Ceiling or floor
+} bleedthrough_t;
+
 static int numfakeplanes = 0;
 static fakegroup_t *fakeplanes = NULL;
 static sector_t **sectors2 = NULL;
+static bleedthrough_t *bleedsectors = NULL;
+static int numbleedsectors = 0;
```
Comment (fabiangreffrath) [inline prboom2/src/gl_missingtexture.c:75]: Where is this ever reset? What happens when entering the next map?

File: prboom2/src/r_bsp.c (hunk 1 above)
Comment (JadingTsunami) [reply prboom2/src/r_bsp.c:43]: No, and I thought of the same later on too. :)

File: prboom2/src/gl_missingtexture.c (hunk 3 above)
Comment (JadingTsunami) [reply prboom2/src/gl_missingtexture.c:207]: Ceiling is a flag with encoding 0=floor, 1=ceiling; this is not my convention as I followed the existing `GetBestFake` implementation.

We could modify both to a better or more explicit indication, or I could use a different convention if it is preferred.

File: prboom2/src/r_bsp.c (hunk 4 above)
Comment (JadingTsunami) [reply prboom2/src/r_bsp.c:705]: Yes, it's nicer this way and does not create strange scoping. Fixed.

File: prboom2/src/gl_missingtexture.c (hunk 6 above)
Comment (JadingTsunami) [reply prboom2/src/gl_missingtexture.c:75]: Good catch, I needed the reset at the top of that loop. Fixed.

File: prboom2/src/gl_missingtexture.c (hunk 5 above)
Comment (JadingTsunami) [reply prboom2/src/gl_missingtexture.c:184]: Ah, yes. I thought this might be nicer for readability but I can do it that way too.

Comment (fabiangreffrath) [reply prboom2/src/gl_missingtexture.c:184]: I see it the other way round, I find it easier to read if this just uses plain `realloc()`.

File: prboom2/src/gl_missingtexture.c [hunk 7]
```diff
@@ -61,21 +61,39 @@ typedef struct
...
   int i, num;
 
+  /* free memory if allocated by previous maps */
+  if (bleedsectors) {
+      free(bleedsectors);
+      numbleedsectors = 0;
+      bleedsectors = NULL;
+  }
+
+
```
Comment (fabiangreffrath) [inline prboom2/src/gl_missingtexture.c:92]: Excess blank line.

File: prboom2/src/gl_missingtexture.c [hunk 8]
```diff
@@ -95,6 +113,44 @@ static void gld_PrepareSectorSpecialEffects(void)
...
+                side0->sector->flags |= MISSING_BOTTOMTEXTURES;
+                gld_RegisterBleedthroughSector(side0->sector,side1->sector,0);
+            }
+
+            if (needs_front_lower && side0->bottomtexture == NO_TEXTURE) {
+                side1->sector->flags |= MISSING_BOTTOMTEXTURES;
+                gld_RegisterBleedthroughSector(side1->sector,side0->sector,0);
+            }
+        }
+
```
Comment (fabiangreffrath) [inline prboom2/src/gl_missingtexture.c:153]: Excess blank line.

File: prboom2/src/gl_missingtexture.c [hunk 9]
```diff
@@ -111,6 +167,58 @@ static void gld_PrepareSectorSpecialEffects(void)
...
+    /* either register the proposed target since it is first,
+     * or check if the new proposed target is a better option
+     * and register it instead */
+
+    if ((bleedsectors[source_idx].target == NULL) ||
+        (bleedsectors[source_idx].target &&
+        ((ceiling && bleedsectors[source_idx].target->ceilingheight > target->ceilingheight) ||
+        (bleedsectors[source_idx].target->floorheight < target->floorheight))))
+        bleedsectors[source_idx].target = target;
```
Comment (fabiangreffrath) [inline prboom2/src/gl_missingtexture.c:210]: This is still hard to read, could you please add braces?

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
Alright, this is mostly about style consistency now, but we are near the finishing line.

File: prboom2/src/gl_missingtexture.c (hunk 5 above)
Comment (JadingTsunami) [reply prboom2/src/gl_missingtexture.c:184]: Sure, addressed this in the latest push.

File: prboom2/src/gl_missingtexture.c (hunk 7 above)
Comment (JadingTsunami) [reply prboom2/src/gl_missingtexture.c:92]: Removed excess blank line.

File: prboom2/src/gl_missingtexture.c (hunk 8 above)
Comment (JadingTsunami) [reply prboom2/src/gl_missingtexture.c:153]: Removed excess blank line.

File: prboom2/src/gl_missingtexture.c (hunk 9 above)
Comment (JadingTsunami) [reply prboom2/src/gl_missingtexture.c:210]: I added braces and expanded the logic in keeping with other files (gl_main.c, am_map.c). See if this is new formatting is preferable.

**Review by fabiangreffrath — APPROVED** [review]

**Comment by fabiangreffrath** [comment]
Is this still WIP as the title suggests?

**Comment by JadingTsunami** [comment]
> Is this still WIP as the title suggests?

I don't have any more work planned on the implementation, but if anyone has feedback I'm happy to incorporate it. I left it as "WIP" in case there was still more to discuss now that everything is updated.

But, I've updated the title to show I am done with the implementation side.

**Pull request merged**